Set the stdout_path and stderr_path the job/task should write to

WilliamMcCumstie opened the pull request:
The stdout and stderr paths are optional inputs to job creation. They direct the daemon to write the job/task output to the specified location.

They support various replacements that can customise the paths on a per node/job basis. See the documentation for a full list of supported replacements. Note the description is in GFM.
```
curl http://<ip>/v0/docs | jq '.definitions.newJob.properties.attributes.properties["stdout-path"].description'
curl http://<ip>/v0/docs | jq '.definitions.newJob.properties.attributes.properties["stderr-path"].description'
```

This PR is based on #12

WilliamMcCumstie committed:
Add a PathGenerator struct to handle the special keys

The paths can have various special keys which need replacing.
This commit documents what they are going to be

WilliamMcCumstie committed:
Implement the shared path attributes

These attributes are determined in a predicable manner and are
shared across all contexts

WilliamMcCumstie committed:
Implement the array path attributes

WilliamMcCumstie committed:
Implement the job ID attribute

WilliamMcCumstie committed:
Allow the pct_* method to render into a string

WilliamMcCumstie committed:
Handle escaping %% without a directive

WilliamMcCumstie committed:
Allow the values to be padded with 0s

WilliamMcCumstie committed:
Define a method for checking validity of a path

The render handles a super set of the valid use cases to
keep its implementation simple

Instead a valid? method has been added for this exact purpose.
It solely validates the `path` and does not require the
job/node/task allocation

WilliamMcCumstie committed:
Only allow numeric chars to have padding

This is done in the validation check which needs to be
preformed up front

WilliamMcCumstie committed:
Use factory_bot in the job_spec

This allows for additional features to Job without constantly
maintaining this spec. Instead the new base feature is added once
to the factory

WilliamMcCumstie committed:
Add the stdout_path and stderr_path attributes to Job

WilliamMcCumstie committed:
Toggle the path default for array jobs

WilliamMcCumstie committed:
Make unescaped % an invalid input

These indicate an unresolved replacement and should be invalid

WilliamMcCumstie committed:
Validate the stdout_path and stderr_path

WilliamMcCumstie committed:
Store the std*_path from the request on the Job

WilliamMcCumstie committed:
Send the rendered paths to the daemon

benarmston commented on lib/flight_scheduler/path_generator.rb:
```diff
@@ -0,0 +1,129 @@
+      # 'n' => 'The relative ID of the node within the job',
+    }
+    ALPHA_KEYS = {
+      'A' => "The ID of the associate job when running an array task, otherwise empty string",
```
Typo: `associated`.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/ae036d1203aa02b303c8571c05d073641644ae47

benarmston commented on lib/flight_scheduler/path_generator.rb:
```diff
@@ -0,0 +1,129 @@
+    }
+    ALPHA_KEYS = {
+      'A' => "The ID of the associate job when running an array task, otherwise empty string",
+      'j' => 'The ID of non array task jobs, otherwise empty string',
```
This should be defined for array tasks too.  It should expand to the array tasks id.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/7101482abd793a47918f45ce1046a6fc9b8b2284

benarmston commented on lib/flight_scheduler/path_generator.rb:
```diff
@@ -0,0 +1,129 @@
+      ' * `%<char>`: All other characters form an invalid replacement'
+    ].join("\n")
+
+    ALL_CHARS = [*ALPHA_KEYS.keys, *NUMERIC_KEYS.keys.join('')]
```
What's the purpose of the `join('')` on `NUMERIC_KEYS`.  We don't have one on `ALPHA_KEYS` and we do have one later on `ALL_CHARS`.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/ae036d1203aa02b303c8571c05d073641644ae47

benarmston commented on lib/flight_scheduler/path_generator.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+
+    def pct_a
+      task ? task.array_index : 0
```
It looks like slurm uses `2**32 - 2` for non-array jobs here.  I'm not sure how consistently it uses that number across platforms and architectures, but it is consistent on the single machine I've tested it.

I can't think of any good reason for us to copy that.  `0` seems like a suitably valid value, and I think it unlikely that anyone is relying on `%a` expanding to `2**32 - 2` for non-array jobs. Probably.

WilliamMcCumstie replied:
Hmm `2**32 -2` looks like an odd number. My guess it's "-1" as an unsigned 32-bit int

WilliamMcCumstie replied:
Yep! It's an algorithmic underflow :laughing:
```
2**32-2 = 0xFFFFFFFE
```

benarmston commented on lib/flight_scheduler/path_generator.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+
+    def pct_A
+      task ? job.id : ''
```
On further inspection of what slurm does, for non-array tasks this should expand to the job's id.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/7101482abd793a47918f45ce1046a6fc9b8b2284

benarmston commented on lib/flight_scheduler/path_generator.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+
+    def pct_j
+      task ? '' : job.id
```
For array tasks, this should expand to the array task's id.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/7101482abd793a47918f45ce1046a6fc9b8b2284

benarmston reviewed (changes requested):
@WilliamMcCumstie this is a good start. I've left a few comments in line for you to look at and there are a couple of general issues to mention.

There is a bug with how standard error is handled.  If standard error and standard output are both set to the same file and a submission script writes to both standard out and standard error one of those is lost.

The reason is that the file is that `::Process.spawn(..., out: 'some/path', err: 'some/path')` will open `some/path` twice if both standard output and standard error are written to.  Both of the following should work correctly.

```ruby
::Process.spawn(..., [:out, :err] => 'some/path')
::Process.spawn(..., out: 'some/path', :err => :out)
```

---

There are also a number of inconsistencies with how slurm handles the path patterns.  These were not mentioned on the trello card as I didn't know about them at the time :-)

* `%` should render as `%`
* `%%%` should render as `%%`
* unsupported formatting characters should render litteraly, e.g., `%q` renders as `%q`.
* Non-numeric formats "support" padding.  E.g., `%5N` should render exactly as `%N` renders.

Ideally, we'd be consistent with slurm on our rendering of these.  However, I'm content to introduce inconsistency if it greatly simplifies the implementation of time to fix.

As far as I can tell, slurm never rejects a job due to an invalid path pattern.  It just might not render as you would expect.

WilliamMcCumstie commented:
Rebased on #18 To resolve various conflicts between the branches. Also allows the `username` to be integrated into the paths

WilliamMcCumstie committed:
Fix various typos

WilliamMcCumstie committed:
Replace %j with the task id for array jobs

WilliamMcCumstie committed:
Ignore the padding directive for alpha chars

WilliamMcCumstie committed:
Confirm that %%% is only escaped once

WilliamMcCumstie committed:
Ensure unrecognised chars and integers are not replaced

WilliamMcCumstie committed:
Return the username from the job

WilliamMcCumstie committed:
Remove validation of stdout_path and stderr_path

These can not be invalid as all inputs are handled. They just may
not do what the user expects

WilliamMcCumstie committed:
Use the default paths instead of empty string

Empty string will almost always a rogue setting in a request
and not the actual intended. Instead the defaults should be used

WilliamMcCumstie commented:
I've updated the replacement so all inputs are now valid. This does mean the `valid?` method needed removing (as it would always return true).

I've also added a few more specs for the specific use cases you mentioned.

benarmston committed:
Keep autoload list ordered

Having the list ordered reduces the changes of merge conflicts compared
to having all new entries added to the end of the list.

benarmston commented:
I've added a commit to keep the autoload list ordered.  Having the list ordered reduces the chances of merge conflicts compared
to having all new entries added to the end of the list.

With that in place this looks good to me.
